golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,421 @@
+	vstoragetypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	snapshots "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+// SnapshotFormat 1 is a proto message containing an artifact name, and the binary artifact data
+const SnapshotFormat = 1
+
+const ExportManifestFilename = "export-manifest.json"
+const ExportDataFilename = "export-data.jsonl"
+const UntrustedExportDataArtifactName = "UNTRUSTED-EXPORT-DATA"
+const UntrustedExportDataFilename = "untrusted-export-data.jsonl"
+const ExportedFilesMode = 0644
+
+var artifactNameSanitizer = regexp.MustCompile(`[^-_.a-zA-Z0-9]`)
+
+type activeSnapshot struct {
+	height        int64
+	startedResult chan error
+	retrieved     bool
+}
+
+type exportManifest struct {
+	BlockHeight uint64      `json:"blockHeight,omitempty"`
+	Data        string      `json:"data,omitempty"`
+	Artifacts   [][2]string `json:"artifacts"`
+}
+
+type SwingStoreExporter interface {
+	ExportSwingStore(ctx sdk.Context) []*vstoragetypes.DataEntry
+}
+
+type SwingsetSnapshotter struct {
+	app            *baseapp.BaseApp
+	logger         log.Logger
+	exporter       SwingStoreExporter
+	blockingSend   func(action vm.Jsonable) (string, error)
+	activeSnapshot *activeSnapshot
+}
+
+type snapshotAction struct {
+	Type        string            `json:"type"` // COSMOS_SNAPSHOT
+	BlockHeight int64             `json:"blockHeight"`
+	Request     string            `json:"request"` // "initiate", "discard", "retrieve", or "restore"
+	Args        []json.RawMessage `json:"args,omitempty"`
+}
+
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, exporter SwingStoreExporter, sendToController func(bool, string) (string, error)) SwingsetSnapshotter {
+	// The sendToController performed by this submodule are non-deterministic.
+	// This submodule will send messages to JS at unpredictable times from a
+	// goroutine. However the JS side does not perform consensus operations when
+	// handling these messages, so it is safe.
+
+	blockingSend := func(action vm.Jsonable) (string, error) {
+		bz, err := json.Marshal(action)
+		if err != nil {
+			return "", err
+		}
+		return sendToController(true, string(bz))
+	}
+
+	return SwingsetSnapshotter{
+		app:            app,
+		logger:         app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "snapshotter"),
+		exporter:       exporter,
+		blockingSend:   blockingSend,
+		activeSnapshot: nil,
+	}
+}
+
+// Synchronously initiates a snapshot for the given height. If a snapshot is
+// already in progress, or if no snapshot manager is configured, this will fail.
+// The snapshot operation is performed in a goroutine, and synchronized with the
+// main thread through the `WaitUntilSnapshotStarted` method.
```

Comment:
[godoc](https://go.dev/doc/comment) nit:
```suggestion
// InitiateSnapshot synchronously initiates a snapshot for the given height.
// If a snapshot is already in progress, or if no snapshot manager is configured,
// this will fail.
// The snapshot operation is performed in a goroutine, and synchronized with the
// main thread through the `WaitUntilSnapshotStarted` method.
```